File: 0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        dp = [-1 for i in range(n+1)]

        if n == 0:
            return [0]
        if n == 1:
            return [0,1]

        dp[0] = 0
        dp[1] = 1
        x = 1
        for i in range(2, n+1):
            if i == 2**(x+1):
                x += 1
            dp[i] = 1 + dp[i-(2**x)]
        
        return dp

File: 0338-counting-bits/0338-counting-bits.cpp
#include <cmath>

class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> dp;
        
        dp.resize(1);
        dp[0] = 0;
        
        if (n == 0) {
            return dp;
        }
        if (n == 1) {
            dp.resize(2);
            dp[1] = 1;
        }
        
        dp.resize(n+1);
        
        dp[0] = 0;
        dp[1] = 1;
        int power = 1;
        
        for (int i = 2; i < n+1; i++) {
            if (i == pow(2, power+1)) {
                power++;
            }
            dp[i] = 1 + dp[i - pow(2, power)];
        }
        
        return dp;
    }
};